hub/core/storage/local.py

```diff
@@ -13,7 +14,137 @@ def __init__(self, root: str):
+        """
+        try:
+            full_path = self._check_is_file(path)
+            file = open(full_path, "rb")
+            return file.read()
+        except DirectoryAtPathException:
+            raise
+        except Exception:
+            raise KeyError
+
+    def __setitem__(self, path: str, value: bytes):
+        """Sets the object present at the path with the value
+
+        Example:
+            local_provider = LocalProvider("/home/ubuntu/Documents/")
+            local_provider["abc.txt"] = b"abcd"
+
+        Args:
+            path (str): the path relative to the root of the provider.
+            value (bytes): the value to be assigned at the path.
+
+        Raises:
+            Exception: If unable to set item due to directory at path or permission or space issues.
+        """
+        try:
+            full_path = self._check_is_file(path)
+            file = open(full_path, "wb")
+            file.write(value)
+        except Exception:
+            raise
+
+    def __delitem__(self, path: str):
+        """Delete the object present at the path.
+
+        Example:
+            local_provider = LocalProvider("/home/ubuntu/Documents/")
+            del local_provider["abc.txt"]
+
+        Args:
+            path (str): the path to the object relative to the root of the provider.
+
+        Raises:
+            KeyError: If an object is not found at the path.
+            DirectoryAtPathException: If a directory is found at the path.
+        """
+        try:
+            full_path = self._check_is_file(path)
+            os.remove(full_path)
+        except DirectoryAtPathException:
+            raise
+        except Exception:
+            raise KeyError
+
+    def __iter__(self):
+        """Generator function that iterates over the keys of the provider.
+
+        Example:
+            local_provider = LocalProvider("/home/ubuntu/Documents/")
+            for my_data in local_provider:
+                pass
+
+        Yields:
+            str: the path of the object that it is iterating over, relative to the root of the provider.
+        """
+        yield from self._list_keys()
+
+    def __len__(self):
+        """Returns the number of files present inside the root of the provider.
+
+        Example:
+            local_provider = LocalProvider("/home/ubuntu/Documents/")
+            len(local_provider)
+
+        Returns:
+            int: the number of files present inside the root.
+        """
+        return len(self._list_keys())
+
+    def _list_keys(self):
+        """Helper function that lists all the objects present at the root of the Provider.
+
+        Returns:
+            list: list of all the objects found at the root of the Provider.
+        """
+        ls = []
+        for r, d, f in os.walk(self.root):
+            for file in f:
+                ls.append(os.path.relpath(os.path.join(r, file), self.root))
+        return ls
+
+    def _check_is_file(self, path: str):
```

Comment:
This is a bad name for this function. It is not clear from the name that this also returns the expanded path, and creates the intermediate directories. It looks like a simple assertion but it isn't.